### DISCUSSION TITLE
How do I store data on the client?

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [zauberzeug/nicegui#872] [Q&A] [answered]
### Question

Ideally the data would persist when he goes from 1 page to another, I presume what I want is cookies or localStorage?

Or should I create a Python dict that uses `nicegui.client.id`'s as keys and put per-client-data in there?

**Comment by rodja** [comment, marked as answer]
Depends on your use-case. `client.id` is unique for each browser-tab as long as you are not using the [auto-index page](https://nicegui.io/documentation#auto-index_page). Using cookies (for [example with the a SessionMiddleware](https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py)) is good for small data. NiceGUI does not provide a simple Python wrapper for localStorage but you can access it simply via JavaScript. Suitable for is for non-sensible larger data (up to 5-10 mb).

**Comment by ed2050** [comment]
@rodja said:
> client.id is unique for each browser-tab as long as you are not using the [auto-index page](https://nicegui.io/documentation#auto-index_page).

1. Is this still accurate?  What conditions cause client.id to change?
2. Is nicegui.client documented anywhere?  I mean the module itself and all its interfaces, not the bits and pieces like 'await client.connected ()` scattered through other documentation.  All I see are [this](https://nicegui.io/documentation/page#pages_with_path_parameters) and [this](https://nicegui.io/documentation/page#wait_for_client_connection).

Thanks.

**Reply by rodja** [reply]
Yes, it is still accurate. Have you any indications that it's not? And yes, more documentation is always good. We are currently thinking about a new structure for https://nicegui.io. Hopefully we find a good place for such "inner workings".

**Reply by ed2050** [reply]
No indications right now.  Simply the above post is 6 months old and a few versions ago.

I'm fine for now.  I'll post what I'm doing here : #2082.
